Use build script to generate binding files

upsuper opened the pull request:
<!-- Please describe your changes on the following line: -->
This is a WIP patch to make rust-bindgen a build time dependency of style component for stylo. This should make things more automatic. I convert majority of `regen.py` to `build_gecko.rs`.

It currently doesn't work on Windows, because of servo/rust-bindgen#271 (it works when I fix the generated file manually, though). I haven't tested other platforms.

It would break servo's CI for geckolib, because:
1. it needs libclang (which is probably easy to solve)
2. it needs `MOZ_OBJDIR` to be passed in so that it can generate bindings

@Manishearth @emilio Do you agree with this change? Do you have suggestion for the issues above?

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [ ] `./mach build -d` does not report any errors
- [ ] `./mach test-tidy` does not report any errors
- [ ] These changes fix #__ (github issue number if applicable).

<!-- Either: -->
- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because _____

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/14244)
<!-- Reviewable:end -->

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/style/Cargo.toml, components/style/gecko_bindings/bindings.rs, components/style/gecko_bindings/structs_debug.rs, components/style/build_gecko.rs, components/style/gecko_bindings/mod.rs, components/style/build.rs, components/style/gecko_bindings/structs_release.rs
 * @emilio: components/style/Cargo.toml, components/style/gecko_bindings/bindings.rs, ports/geckolib/Cargo.lock, components/style/gecko_bindings/structs_debug.rs, components/style/build_gecko.rs, components/style/gecko_bindings/mod.rs, components/style/build.rs, components/style/gecko_bindings/structs_release.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!
* These commits modify style code, but no tests are modified. Please consider adding a test!

Wafflespeanut commented:
While automating the bindings generation is a wonderful idea, I'm ambivalent about this for practical use, especially because of we'll be needing a gecko build for a simple `geckolib` build. We could still pass an argument to `./mach build-geckolib` for setting the `MOZ_OBJDIR`, but I feel this would be an unnecessary dependency while refactoring existing glue code or the FFI code, for example.

Manishearth commented:
> While automating the bindings generation is a wonderful idea, I'm ambivalent about putting this for practical use, especially because of we'll be needing a gecko build for a simple geckolib build

Yeah. To me, build-time bindgen only makes sense when we have a merged repo. Until we have the double-CI workflow, geckolib _must_ build on its own so that we have some reasonable level of CI assurance.

(We can, of course, make build time bindgen an option that only runs when the MOZ_OBJDIR env var is present)

Wafflespeanut commented:
> (We can, of course, make build time bindgen an option that only runs when the MOZ_OBJDIR env var is present. Folks wanting to update the in-tree structs file can run build time bindgen and copy the generated files)

Yes, we should consider this. We can make use of `mach` to help ease the workflow of the people who generate the bindings (@emilio, @Manishearth or @heycam) :smile:

upsuper commented:
Making build time bindgen run only when MOZ_OBJDIR is present would add burden of maintaining sync between regen.py and build_gecko.rs. Also people would still need to check in binding files which they may forget when local have things done automatically.

Another option is that we drop this and continue using regen.py. Just let gecko's mach invoke that script explicitly. That way, we may need to figure out an approach to teach backends understand the dependency. That also means gecko need to vendor rust-bindgen manually.

emilio commented:
This is great in my opinion, and it's probably where we want to go in the long term anyway, so +1 from me, though breaking the build in servo's CI is definitely a problem that we should probably fix.

upsuper commented:
We may remove regen.py but still keep binding files in tree, and make build script use the in-tree files when `MOZ_OBJDIR` is not available. Then people can manually copy the binding files from target dir to in-tree location. Locating output dir may take some time, but I guess that wouldn't be more troublesome than manually running regen.py.

Manishearth commented:
> Making build time bindgen run only when MOZ_OBJDIR is present would add burden of maintaining sync between regen.py and build_gecko.rs.

Not necessarily? I said we only run it when MOZ_OBJDIR is present, not that we should keep regen.py. Like I said, "Folks wanting to update the in-tree structs file can run build time bindgen and copy the generated files"

There's only a slight issue with this -- in my workflow when working on new things I usually sync and check the diff (since sometimes I mess up) and we no longer get to do that. Oh well.

bholley commented:
@indygreg is currently working on [1] as a stopgap to autolander, and is doing something related to automatic binding generation. I'll let him weigh on on whether this makes things better or worse in the mean time.

[1] https://bugzilla.mozilla.org/show_bug.cgi?id=1317525

Manishearth commented on components/style/build.rs:
```diff
@@ -62,3 +71,11 @@ fn main() {
+
+fn main() {
+    println!("cargo:rerun-if-changed=build.rs");
+    generate_properties();
```
We need to also be able to trigger a rebuild if the contents of the objdir change, which is a bit harder to do here.

Perhaps make the build script unconditional and have it detect this?

upsuper replied:
That is done in build_gecko.rs which recursively adds header files which it may depend on to `cargo:rerun-if-changed`.

Manishearth commented on components/style/gecko_bindings/bindings.rs:
```diff
@@ -1,1211 +0,0 @@
-/* automatically generated by rust-bindgen */
```
please don't remove these. Commit the newly generated versions.

Manishearth reviewed (changes requested):
See if you can make this optional.

Also, r?@emilio on the new bindings script? You're more familiar with the bindgen args. It overall looks correct to me, but I may have missed something.

froydnj commented:
> Another option is that we drop this and continue using regen.py. Just let gecko's mach invoke that script explicitly. That way, we may need to figure out an approach to teach backends understand the dependency. That also means gecko need to vendor rust-bindgen manually.

Keeping `regen.py` requires a lot more plumbing for integration on the Gecko side; Gecko's build system strongly prefers that you generate files with Python scripts, not with Rust binaries.  Teaching it to build Rust binaries to invoke them ("natively" or via Python script wrapper) would be some work.  It would be preferable if all this stuff was invoked via Rust build scripts.

I'm not up to speed on the terminology (geckolib?), but it sounded to me like an automatic build-time solution + semi-manual generation via `mach build-geckolib` would work OK?  Please feel free to correct my ignorance. 😄

Manishearth commented:
> I'm not up to speed on the terminology (geckolib?)

geckolib is the name of the stylo library built from the servo repo that gets linked into Firefox. It can currently be built independently of a Firefox build. Until the CI is all there, I'd like it to stay that way.

upsuper commented:
Now the blockers are servo/rust-bindgen#272 and servo/rust-bindgen#274.

upsuper commented:
It is almost done. Now build-time bindgen is optional. It would use the in-tree version if `MOZ_OBJDIR` is not available. There is still one problem that rust-bindgen has duplicate versions for package aster, which we need to resolve, otherwise test-tidy would complain.

upsuper commented:
Travis CI's libclang version is too old... That is 3.4... rust-bindgen requires at least 3.8. Is there a chance we can upgrade the system we use there?

upsuper commented:
The test-tidy failure is because of duplicate versions of aster. That is an issue of rust-bindgen which I have pr for: servo/rust-bindgen#275.

nox commented:
Just saw this. Can we avoid the dependency on syntex_syntax?

emilio commented:
You can look at how to install modern libclang on Travis on the bindgen CI script: https://github.com/servo/rust-bindgen/blob/master/ci/before_install.sh

bors-servo commented:
:umbrella: The latest upstream changes (presumably #14245) made this pull request unmergeable. Please resolve the merge conflicts.

bholley commented:
@upsuper, how close are we on this? I think it's probably necessary to make our auto-incubator stuff work (currently the stylo builds are all failing from binding issues).

upsuper commented:
@bholley it has almost completed. I already had a local patch which can use the in-tree bindings if the needed environment variable is not passed in. The remaining issue for this pr itself is to include libclang in the CI environment so that we don't break the CI. I can probably just take what rust-bindgen does.

I was waiting for [bug 1318981](https://bugzilla.mozilla.org/show_bug.cgi?id=1318981) because that's a prerequisite for this to work as expected. And I am currently waiting for @emilio's work to fix rust-bindgen, because I expect that would also change the original regen.py script quite a bit.

If the change to regen.py isn't coming soonish but this change is critical, I can try to complete this first, so that further changes would go to the new build script directly. I didn't realize that this has a very high priority.

bholley commented:
Assuming this is going to be optional, we don't need to wait on Travis changes, right? I thought the idea was that the Travis/Homu tests would still used the checked-in bindings, and that build-time bindgen would be used on incubator.

upsuper commented:
This is going to be optional at runtime of the build script, but to build the build script, we need libclang to present. clang-sys supports loading libclang at runtime, maybe we can make rust-bindgen support that as well. (But any change to rust-bindgen required would also mean we may be blocked on upgrading it.)

That also reminds me, does the Gecko CI have libclang right now?

bholley commented:
> That also reminds me, does the Gecko CI have libclang right now?

Good question - @indygreg?

upsuper commented:
> This is going to be optional at runtime of the build script, but to build the build script, we need libclang to present.

Oh, actually we can hide build-time bindgen behind a feature, and only enable that feature in stylo build, but not normal geckolib build, so that we don't need to depend on it in Servo's CI.

froydnj commented:
> That also reminds me, does the Gecko CI have libclang right now?

The files exist in tooltool, but I am unsure of where to indicate they
should be downloaded, or what environment variables we need to setup for
cargo.

upsuper commented:
> The files exist in tooltool, but I am unsure of where to indicate they should be downloaded, or what environment variables we need to setup for cargo.

It is 3.9 or 3.8?

We need it to be installed in some standard location, or its location is passed in via `LIBCLANG_PATH` environment.

clang-sys searches the following paths if no environment is provided on Linux:
> /usr/lib*
> /usr/lib*/*
> /usr/lib*/*/*
> /usr/local/lib*
> /usr/local/lib*/*
> /usr/local/lib*/*/*
> /usr/local/llvm*/lib

upsuper committed:
Do build-time bindgen

Majority of build_gecko.rs is just the straightforward conversion from
regen.py. There are two differences that:
1. Side in whitelist is changed to mozilla::Side
2. std::atomic__My_base is added to opaque types for Windows

upsuper committed:
Make regen_atoms.py work by itself

upsuper committed:
Remove script and doc no longer relevant

upsuper commented:
I would say it is done now.

r? @emilio

upsuper commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit 4b24a12 with merge 3947930...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/1358)

bholley commented:
@bors-servo retry infra

bors-servo commented:
:hourglass: Trying commit 4b24a12 with merge 9a3f282...

emilio commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 4b24a12 has been approved by `emilio`

<!-- @bors-servo r=emilio 4b24a121d47a3d26c72a258389a1a9a1e44e5f68 -->

KiChjang commented:
@bors-servo r=emilio

bors-servo commented:
:pushpin: Commit d5e59eb has been approved by `emilio`

<!-- @bors-servo r=emilio d5e59eb0c114559f93a48e951f3aaf8d0bf5f1b9 -->

bors-servo commented:
:lock: Merge conflict

bors-servo commented:
:umbrella: The latest upstream changes (presumably #14535) made this pull request unmergeable. Please resolve the merge conflicts.

KiChjang commented:
@bors-servo delegate=upsuper

bors-servo commented:
:v: @upsuper can now approve this pull request

upsuper commented:
@bors-servo r=emilio

bors-servo commented:
:pushpin: Commit 6e78325 has been approved by `emilio`

<!-- @bors-servo r=emilio 6e783258667d093ad7d2b7f5087c49289ed17855 -->

bors-servo commented:
:hourglass: Testing commit 6e78325 with merge 8ed3caa...

bors-servo commented:
:broken_heart: Test failed - [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/4640)

upsuper commented:
@bors-servo retry infra

bors-servo commented:
:hourglass: Testing commit 6e78325 with merge 23d864c...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/1367)

upsuper commented:
@bors-servo retry infra

bors-servo commented:
:hourglass: Testing commit 6e78325 with merge 872ec89...

bors-servo commented:
:sunny: Test successful - [arm32](http://build.servo.org/builders/arm32/builds/4575), [arm64](http://build.servo.org/builders/arm64/builds/4584), [linux-dev](http://build.servo.org/builders/linux-dev/builds/4665), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/1343), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/1361), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/4641), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/4665), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/1354), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/1368), [windows-dev](http://build.servo.org/builders/windows-dev/builds/2610)
